crates/feldera-storage/src/file/writer.rs

```diff
@@ -0,0 +1,1031 @@
+
+    /// Minimum branching factor.  This controls the minimum number of data
+    /// items in a data block and the minimum number of child nodes in an index
+    /// block.
+    ///
+    /// Increasing the branching factor reduces the number of nodes that must be
+    /// read to find a particular value.  It also increases the memory required
+    /// to read each block.
+    ///
+    /// The branching factor is not an important consideration for small data
+    /// values, because our 4-kB (or larger) minimum data and index block size
+    /// means that the minimum branching factor will be high.  For example, over
+    /// 100 32-byte values fit in a 4-kB block, even considering overhead.
+    ///
+    /// The branching factor is more important for large values.  Suppose only
+    /// a single value fits into a data block.  The index block that refers to
+    /// it reproduces the first value in each data block, which in turn makes
+    /// it likely that index block only fits a single child, which is
+    /// pathological and silly.
+    pub min_branch: usize,
+}
+
+impl Default for Parameters {
+    fn default() -> Self {
+        Self {
+            min_data_block: 8192,
+            min_index_block: 8192,
+            min_branch: 32,
+        }
+    }
+}
+
+trait IntoBlock {
+    fn into_block(self) -> AlignedVec;
+    fn overwrite_head(&self, dst: &mut AlignedVec)
+    where
+        Self: FixedLen;
+}
+
+impl<B> IntoBlock for B
+where
+    B: for<'a> BinWrite<Args<'a> = ()>,
+{
+    fn into_block(self) -> AlignedVec {
+        let mut block = NoSeek::new(AlignedVec::with_capacity(4096));
+        self.write_le(&mut block).unwrap();
+        block.into_inner()
+    }
+
+    fn overwrite_head(&self, dst: &mut AlignedVec)
+    where
+        Self: FixedLen,
+    {
+        let mut writer = Cursor::new(dst.as_mut_slice());
+        self.write_le(&mut writer).unwrap();
+        debug_assert_eq!(writer.position(), <Self as FixedLen>::LEN as u64);
+    }
+}
+
+struct Writer<W>
+where
+    W: Write + Seek,
+{
+    writer: W,
+    cws: Vec<ColumnWriter>,
+    finished_columns: Vec<FileTrailerColumn>,
+}
+
+struct ColumnWriter {
+    parameters: Rc<Parameters>,
+    column_index: usize,
+    rows: Range<u64>,
+    data_block: DataBlockBuilder,
+    index_blocks: Vec<IndexBlockBuilder>,
+}
+
+impl ColumnWriter {
+    fn new(parameters: &Rc<Parameters>, column_index: usize) -> Self {
+        ColumnWriter {
+            parameters: parameters.clone(),
+            column_index,
+            rows: 0..0,
+            data_block: DataBlockBuilder::new(parameters),
+            index_blocks: Vec::new(),
+        }
+    }
+
+    fn take_rows(&mut self) -> Range<u64> {
+        let end = self.rows.end;
+        replace(&mut self.rows, end..end)
+    }
+
+    fn finish<W, K, A>(&mut self, writer: &mut W) -> IoResult<FileTrailerColumn>
+    where
+        W: Write + Seek,
+        K: Rkyv,
+        A: Rkyv,
+    {
+        // Flush data.
```

Comment:
I guess they work together. `take().build()` takes the block under construction and puts it in the form that it can be written to disk, and write_data_block does the write and adds it to the index.